ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java

```diff
@@ -201,11 +203,37 @@ private Response makeGetRequest(String path) {
     public QueryStream(Response response) {
       this.response = response;
 
+      this.closed = false;
       this.objectMapper = new ObjectMapper();
-      this.responseScanner = new Scanner((InputStream) response.getEntity(), StandardCharsets.UTF_8.name());
+      InputStreamReader isr = new InputStreamReader((InputStream)response.getEntity(), StandardCharsets.UTF_8);
+      Object lk = this;
+      String errMsg = closedIllegalStateMsg("hasNext()");
+      this.responseScanner = new Scanner(new Readable() {
+        public int read(CharBuffer buf) throws IOException {
+          int wait = 1;
+          // poll the input stream's readiness between interruptable sleeps
+          // this ensures we cannot block indefinitely on read()
+          while (true) {
+            if (closed) throw new IllegalStateException(errMsg);
```

Comment:
why are we doing this check outside the `synchronized`? if `closed` is not volatile there is a chance this thread won't see the change. I think it would be better to move into the `synchronized` and removed the double check

Reply:
Thanks. The check is there so that we exit even if a line is ready after being interrupted. I'll move it into the synchronized block.